mdtraj/core/topology.py

```diff
@@ -114,16 +121,15 @@ def _topology_from_subset(topology, atom_indices):
                     )
                     old_atom_to_new_atom[atom] = newAtom
 
-    bondsiter = topology.bonds
+    bondsiter: Iterator[Bond] | list[Bond] = topology.bonds
     if not hasattr(bondsiter, "__iter__"):
-        bondsiter = bondsiter()
+        bondsiter = list(bondsiter)
```

Comment:
Making this a `list` seems like a behavior change, and I'm leaning towards not doing it. Do we have any insight into when this condition might be used? (what would `topology.bonds` have to be to go through here?

If the tests are not hitting this line, it's possible that we might not see the effect of a change through tests. We may want to check the codecov :)

Reply:
I checked the codecov and line 126 is not covered by the tests although while I was reading it, this part of the code is just trying to ensure that `bondsiiter` is an iterator  and mypy is right about `bondsiter` is not a callable. I think after adding all the type annotations, it should be already gaurenteed as an iterator so this check may not be necessary.

Reply:
I'm not sure what kind of an iterator this is, but potentially unwrapping a generator is an undesirable change

Reply:
Definitely agree! We do not want to unwrap a generator. For the typing on line 124, it would be *best* if we could figure out what was the intention of this `if` (when did they expect it to go through this?)

However, lacking that what we can tell is that if we go through the if statement `bondsiter` is a callable that gives us something we can iterate through (as seen later in the `for` loop.

So I think the typing on like 124 would be

```
bondsiter: Iterable[Bond] | Callable[[], Iterable[Bond]] = topology.bonds
```

And to change the portion under the `if` statement back to the original code.

Reply:
@janash I agree this is the better solution. I'll change it accordingly. Thank you!